bot/cogs/doc/cog.py

```diff
@@ -52,18 +53,83 @@
 NOT_FOUND_DELETE_DELAY = RedirectOutput.delete_delay
 
 
+class TODO_PLACEHOLDER:
+    __slots__ = ("queue", "item_events", "started", "fetch_lock")
+    results_dict: Dict[str, discord.Embed] = None
+
+    def __init__(self):
+        self.queue: List[DocItem] = list()
+        self.fetch_lock = asyncio.Lock()
+        self.started = False
+        self.item_events: Dict[DocItem, asyncio.Event] = {}
+
+    @classmethod
+    def set_results_dict(cls, results_dict: Dict[str, discord.Embed]):
+        """
+        Set dict for parse results for all instances.
+
+        Must be set before any parsing is attempted.
+        """
+        cls.results_dict = results_dict
+
+    def add_item(self, item):
+        """Add `item` into the parse queue."""
+        self.queue.append(item)
+
+    def put_to_front(self, item):
+        """Move `item` to the front of the parse queue."""
+        self.queue.remove(item)
+        self.queue.append(item)
+
+    async def wait_for_item(self, symbol_info, soup: Optional[BeautifulSoup] = None):
+        """
+        Block until `symbol_info` is parsed and assigned to `cls.results_dict`.
+
+        When not started, `soup` must be provided to start the parsing task;
+        after the initial request, providing the soup is unnecessary.
+        """
+        self.put_to_front(symbol_info)
+        item_event = asyncio.Event()
+        # First request
+        if not self.item_events:
+            log.info(f"Started {symbol_info.base_url + symbol_info.relative_url_path}")
+            self.item_events[symbol_info] = item_event
+            asyncio.create_task(self.parse_items(soup))
+        else:
+            self.item_events[symbol_info] = item_event
+        await item_event.wait()
+
+    async def parse_items(self, soup):
+        """Parse all item from `soup` and assign their result embeds into `cls.results_dict`."""
+        while self.queue:
+            item = self.queue.pop()
+            embed = discord.Embed(
+                title=discord.utils.escape_markdown(item.name),
+                url=item.url,
+                description=get_symbol_markdown(soup, item)
+            )
```

Comment:
Returning an embed feels like piling on too much responsibility to one class. Let's strip it down and keep to its minimum - returning the markdown description.